include/dali/core/geom/mat.h

```diff
@@ -539,6 +540,82 @@ std::ostream &operator<<(std::ostream &os, const dali::mat<rows, cols, T> &m) {
   return os;
 }
 
+template <int n, int m, typename T>
+DALI_HOST_DEV
+inline void solve_gauss(mat<n, n, T> &A, mat<n, m, T> &B) {
+  for (int v = 0; v < n; v++) {
+    T max = std::abs(A(v, v));
+    int maxr = v;
+    for (int i = v + 1; i < n; i++) {
+      T q = std::abs(A(i, v));
+      if (q > max) {
+        max = q;
+        maxr = i;
+      }
+    }
+    if (!max) {
+      #ifdef __CUDA_ARCH__
+        continue;
+      #else
+        throw std::range_error("Cannot solve the system with a singlular matrix.");
+      #endif
+    }
+    if (maxr != v) {
+      for (int j = 0; j < n; j++)
+        cuda_swap(A(v, j), A(maxr, j));
+      for (int j = 0; j < m; j++)
+        cuda_swap(B(v, j), B(maxr, j));
+    }
+    T x = A(v, v);
+    x = 1/x;
+    A(v, v) = 1;
+    for (int j = v + 1; j < n; j++)
+      A(v, j) *= x;
+    for (int j = 0; j < n; j++)
+      B(v, j) *= x;
+    for (int i = 0; i < n; i++) {
+       if (i == v)
+        continue;
+#ifndef __CUDA_ARCH__
+      using std::fma;
+#endif
+      T c = -A(i, v);
+      A(i, v) = 0;
+      for (int j = v + 1; j < n; j++)
+        A(i, j) = fma(c, A(v, j), A(i, j));
+      for (int j = 0; j < m; j++)
+        B(i, j) = fma(c, B(v, j), B(i, j));
+    }
+  }
+}
+
+template <int n, typename T, typename U = decltype(1.0f + T())>
+DALI_HOST_DEV
+inline mat<n, n, U> inverse(mat<n, n, T> A) {
+  mat<n, n, U> O = 1;
+  if (std::is_floating_point<T>::value) {
+    solve_gauss(A, O);
+  } else {
+    mat<n, n, U> tmp = A;
+    solve_gauss(tmp, O);
+  }
+  return O;
+}
+
+template<typename T, typename U = decltype(1.0f + T())>
+DALI_HOST_DEV
+inline mat<2, 2, U> inverse(mat<2, 2, T> A) {
+  auto det = A(0, 0)*A(1, 1) - A(0, 1)*A(1, 0);
+  if (!det) {
+    #ifdef __CUDA_ARCH__
+      return mat<2, 2, U>::eye();
```

Comment:
```suggestion
      return 1;
```
nitpick: You were using this syntax above. Maybe unify?

Reply:
done